### PR TITLE
[Clang] Add captures to the instantiation scope for noexcept specifiers

### LINKED ISSUE #95735: [Clang]: clang rejects valid codes when it trying capturing something never occuring in lambda body
```c+++

struct A {
  void operator()(int) {}
};
inline constexpr struct {
  constexpr auto operator()(A fn) const
  {
    return [f{fn}](auto &&tpl) 
        noexcept(noexcept(&f))
    { return true; };
  }
} error;

inline constexpr struct {
  constexpr auto operator()(A fn) const
  {
    return [f{fn}](auto &&tpl) 
        // noexcept(noexcept(&f))
    { return true; };
  }
} no_error;



int main() {
  (void)error( A{} )(1);    // $1
  (void)no_error( A{} )(1); // $2

}
```

- clang trunk with flag `c++26`
- https://godbolt.org/z/1EK6qf8Th
  - see `no_error` and `error` function object, if I delete the `noexcept` clause then everything is OK.
  - clang tells me to capture `fn`, which is weird (since `fn` is never used in function body), but if I do so, clang accepts it.
  - if I change `f` inside the `noexcept` to `fn`, then again, everything would be fine.

clang rejects $1 it with:

```
<source>:8:15: error: variable 'fn' cannot be implicitly captured in a lambda with no capture-default specified
    8 |     return [f{fn}](auto &&tpl) 
      |               ^
<source>:8:12: note: in instantiation of exception specification for 'operator()<int>' requested here
    8 |     return [f{fn}](auto &&tpl) 
      |            ^
<source>:26:21: note: in instantiation of function template specialization '(anonymous struct)::operator()(A)::(anonymous class)::operator()<int>' requested here
   26 |   (void)error( A{} )(1);
      |                     ^
<source>:6:31: note: 'fn' declared here
    6 |   constexpr auto operator()(A fn) const
      |                               ^
<source>:8:12: note: lambda expression begins here
    8 |     return [f{fn}](auto &&tpl) 
      |            ^
<source>:8:12: note: capture 'fn' by value
    8 |     return [f{fn}](auto &&tpl) 
      |            ^
      |            fn
<source>:8:12: note: capture 'fn' by reference
    8 |     return [f{fn}](auto &&tpl) 
      |            ^
      |            &fn
1 error generated.
```

but accepts $2. 

## Expected behavior

Accepts both of them.

### PR DESCRIPTION
**Pull request by zyn0217**
The noexcept specifiers of dependent lambdas would be transformed and rebuilt, where the map of instantiation should also contain captured variables in case they are used from the noexcept specifier.

I also uncovered another assertion failure while at it. However, I decided to leave it as-is because 1) that doesn't appear to be the case in the release version and 2) fixing that might lead to ABI breakage. Anyhow, the case has been added to the test comment.

Fixes https://github.com/llvm/llvm-project/issues/95735

### CODE REVIEW

File: clang/test/SemaTemplate/generic-lambda.cpp [hunk 1]
```diff
@@ -60,3 +60,26 @@ template<class T1> C1<X<X<T1>>> auto t3() {
...
+
+int foo(auto... fn) {
+  // FIXME: This one hits the assertion "if the exception specification is dependent,
+  // then the noexcept expression should be value-dependent" in the constructor of
+  // FunctionProtoType.
+  // One possible solution is to update Sema::canThrow() to consider expressions
+  // (e.g. DeclRefExpr/FunctionParmPackExpr) involving unexpanded parameters as Dependent.
+  // This would effectively add an extra value-dependent flag to the noexcept expression.
+  // However, I'm afraid that would also cause ABI breakage.
+  // [...f = fn](auto tpl) noexcept(noexcept(f)) { return 0; }(0);
```
Comment (cor3ntin) [inline clang/test/SemaTemplate/generic-lambda.cpp:78]: Are we missing an unexpanded pack check here?

**Review by cor3ntin — APPROVED** [review]
LGTM, thanks!
I think your fixme can be resolved in a follow up PR

**Comment by zyn0217** [comment]
Thanks for the review.

File: clang/test/SemaTemplate/generic-lambda.cpp (hunk 1 above)
Comment (zyn0217) [reply clang/test/SemaTemplate/generic-lambda.cpp:78]: We have that check: with a non-assertion build, we would correctly diagnose against the unexpanded pack here.
However, that check happens after we form a FunctionProtoType, where an assertion would have been encountered.

**Pull request merged**